core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala

```
@@ -0,0 +1,105 @@
+  }
+
+  private def getToken: String = {
+    val cachedToken = tokenCache.get("token")
+    if (isTokenInvalid) {
+      val resp = scalaj.http.Http(tokenUrl)
+        .postForm(Seq(
+          "grant_type" -> grantType,
+          "client_id" -> clientId,
+          "client_secret" -> clientSecret,
+          "scope" -> scope))
+        .asString
+
+      val jsonResp = scala.util.parsing.json.JSON.parseFull(resp.body)
+      jsonResp match {
+        case Some(map: scala.collection.immutable.Map[String, Any]) =>
+          val token = map.getOrElse("access_token", "").toString
+          if (token.nonEmpty) {
+            val expirationTime = Instant.now().plus(Duration.ofSeconds(map.getOrElse("expires_in", 0).toString.toDouble.toInt))
+            val newToken = Token(token, expirationTime)
+            tokenCache.put("token", newToken)
+          }
+          token
+        case _ =>
+          throw new RuntimeException(failureMessage)
+      }
+    } else {
+        cachedToken.getOrElse(throw new RuntimeException(failureMessage)).tokenValue
+    }
+  }
+
+  override def createRequest(httpRequest: HttpRequest, authentication: scala.collection.immutable.Map[String, String]): HttpRequest = {
+        val token = getToken
+        httpRequest.header("Authorization", s"Bearer $token")
+  }
+}
+
+
+object AuthenticationFactory extends Logging {
+  def createAuthentication(authentication: scala.collection.immutable.Map[String, String]): Authentication = {
+    if (authentication.isDefinedAt("mode") && authentication.isDefinedAt("grantType")) {
+      authentication("mode") match {
+          case "enabled" => authentication("grantType") match {
+                                  case "client_credentials" => ClientCredentialsAuthentication(authentication)
+                                  case _ =>
+                                    logInfo(s"$authentication('grantType') not implemented")
+                                    NoAuthentication(authentication)
```

too much indentation

---

When `authentication.mode` is `enabled` and some mandatory parameter is missing, we should throw an exception.